### PR TITLE
gh-103333: Pickle the keyword attributes of AttributeError

### PR DESCRIPTION
**Pull request by csm10495**
# gh-103333: Pickle the keyword attributes of AttributeError

This should fix the case where pickling/unpickling an AttributeError loses the name/obj fields. 

If this looks ok, I'd be happy to add a unit test, blurb, etc.

### CODE REVIEW

**Comment by Eclips4** [comment]
Hello!
There's would be nice to have a tests for it case.

**Comment by csm10495** [comment]
Sure, I'll write tests, update blurb once I know that this approach is likely to be accepted :)

**Comment by gaogaotiantian** [comment]
It won't be a rare case when the object is not picklable right? Won't that make a picklable Exception non-picklable with this change?

**Comment by csm10495** [comment]
In theory yeah, if the object isn't picklable it would not work well here. Though this format follows in line with how other exceptions are made to be picklable (like OSError). Is there a way to somehow check if the obj, etc. are picklable and set them in that case (and if not, just leave as None (as it is now)).

**Comment by csm10495** [comment]
I wonder if there should be some sort of pickle mode that pickles all it can and for the fields it couldn't pickle: set to `None` or something else. Maybe even it could recurse and make a new object with 'as many fields as it could serialize' again.

**Comment by gaogaotiantian** [comment]
I think `OSError` guarantees that all the arguments are picklable(at least when used correctly). I'm not 100% sure but I believe that some of the standard libraries rely on the fact that all the built-in Exceptions are picklable - `multiprocesing` is one of them if I remember correctly. Not sure if making a built-in Exception unpicklable is an option here. I think @iritkatriel has been working on Exceptions a lot and must have a much better understanding than I do.

**Comment by csm10495** [comment]
The original issue came from a user using AttributeError via multiprocessing and the .name attribute was lost via the pickle. .. maybe a middle ground is to just serialize .name and postpone the conversation on .obj?

**Comment by gaogaotiantian** [comment]
Having an Exception that loses attributes after pickling is definitely better than having one that is not picklable sometimes. Supporting pickling the `name` attribute seems reasonable to me, but the core devs are the final decision makers.

**Comment by csm10495** [comment]
Updated PR to not pickle obj. Also updated a test to hit this change as well.

**Comment by iritkatriel** [comment]
CC @serhiy-storchaka re pickle.

**Comment by csm10495** [comment]
@iritkatriel @serhiy-storchaka any chance for a merge or feedback?

**Comment by csm10495** [comment]
@iritkatriel @serhiy-storchaka please provide feedback and/merge. Thanks.

File: Lib/test/test_exceptions.py
Comment (gpshead) [inline Lib/test/test_exceptions.py]: modern code style nit: upgrade this to an f-string while you're editing it.

File: Objects/exceptions.c
Comment (gpshead) [inline Objects/exceptions.c]: The comment text is good to keep, but don't include commented out code. It won't be maintained.

File: Objects/exceptions.c
Comment (gpshead) [inline Objects/exceptions.c]: why is `self->obj` in this condition given we're not going to pickle it?

**Review by gpshead — CHANGES_REQUESTED** [review]

**Comment by gpshead** [comment]
I left some comments, but I don't feel confident judging what changes to something that gets pickled are valid compatibility wise as I generally avoid use of pickle outside of tightly controlled scenarios.

Will there be any compatibility issue with an AttributeError pickled by this PR being loaded in a Python process running an older Python version without this PR?

**Comment by csm10495** [comment]
@gpshead I updated based off your feedback. In terms of pickle-ability it works out well:

```
Python 3.12.0a7+ (heads/attributeerror_pickling-dirty:0ee8b8cb, May 10 2023, 19:09:08) [MSC v.1932 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> a = AttributeError(name='test123')
>>> print(a.name)
test123
>>> import pickle
>>> pickle.dumps(a)
b'\x80\x04\x95@\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x0eAttributeError\x94\x93\x94)R\x94}\x94(\x8c\x04name\x94\x8c\x07test123\x94\x8c\x04args\x94)ub.'
>>> pickle.loads(_).name
'test123'
>>>
```

Then in Python 3.11 I try to use the dumped pickle bytes:

```
Python 3.11.0 (main, Oct 24 2022, 18:26:48) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> a = pickle.loads(b'\x80\x04\x95@\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x0eAttributeError\x94\x93\x94)R
>>> a.name
'test123'
>>> pickle.dumps(AttributeError(name='hello'))
b'\x80\x04\x95"\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x0eAttributeError\x94\x93\x94)R\x94.'
```

Then back in 3.12.0a7, using that pickle data works, albeit without the name field:

```
>>> pickle.loads(b'\x80\x04\x95"\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x0eAttributeError\x94\x93\x94)R\x94.')
AttributeError()
>>> pickle.loads(b'\x80\x04\x95"\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x0eAttributeError\x94\x93\x94)R\x94.').name
```

Does this seem ok for merge now?

File: Objects/exceptions.c
Comment (erlend-aasland) [inline Objects/exceptions.c]:
```suggestion
        if (dict == NULL) {
            return NULL;
        }
```

File: Objects/exceptions.c
Comment (erlend-aasland) [inline Objects/exceptions.c]:
```suggestion
    Py_RETURN_NONE;
```

File: Objects/exceptions.c
Comment (erlend-aasland) [inline Objects/exceptions.c]:
```suggestion
    if (state == NULL) {
        return NULL;
    }
```

**Review by erlend-aasland — COMMENTED** [review]
PEP-7 nits

**Comment by csm10495** [comment]
Fixed nits. Thanks!

**Review by gpshead — APPROVED** [review]

**Comment by iritkatriel** [comment]
There is a refleak now in this test, could it be this PR?

```
% ./python.exe -m test -R3:3 test_exceptions -m testAttributes 
Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 3.46 Run tests sequentially
0:00:00 load avg: 3.46 [1/1] test_exceptions
beginning 6 repetitions
123456
......
test_exceptions leaked [18, 18, 18] references, sum=54
test_exceptions leaked [12, 12, 12] memory blocks, sum=36
test_exceptions failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_exceptions

Total duration: 256 ms
Tests result: FAILURE

```

**Comment by iritkatriel** [comment]
See https://github.com/python/cpython/issues/104454

**Pull request merged**
